Change kube hairpin configuration settings and defaults

juanluisvaladas opened the pull request:
## Description

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->


Enable hairpin for kube-router by default

We agreed that the most logical decision was to enable hairpin by default. The old configuration doesn't reflect our real needs, for this reason we decided to deprecate the old configuration and add a new configuration. This commit handles all these details.

Also I spotted a bug related to reconciling, since I noticed it in one of the files which I spotted and it's just a one liner I slipped the commit in this PR.

Fixes #1953

## Type of change

<!-- check the related options -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

## How Has This Been Tested?

- [x] Manual test
- [x] Auto test added

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

## Checklist:

- [x] My code follows the style [guidelines](https://github.com/k0sproject/k0s/blob/main/docs/contributors/overview.md) of this project 
- [x] My commit messages are [signed-off](https://github.com/k0sproject/k0s/blob/main/docs/contributors/github_workflow.md)
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

juanluisvaladas commented:
@hercherf  @mtb-xt since you commented on #1953 I'd like your feedback as well please.

twz123 commented on pkg/apis/k0s.k0sproject.io/v1beta1/kuberouter.go:
```diff
@@ -24,19 +24,32 @@ type KubeRouter struct {
 	PeerRouterIPs string `json:"peerRouterIPs"`
 }
 
+type Hairpin string
+
+const (
+	HairpinEnabled  Hairpin = "enabled"
+	HairpinAllowed  Hairpin = "allowed"
+	HairpinDisabled Hairpin = "disabled"
+	// Necessary for backwards compatibility with HairpinMode
+	HairpinUndefined Hairpin = ""
+)
+
```
I'd uppercase the values, and I'd add an enum hint for kubebuilder.

```suggestion
// +kubebuilder:validation:Enum=Enabled;Allowed;Disabled
type HairpinMode string

const (
	HairpinEnabled  HairpinMode = "Enabled"
	HairpinAllowed  HairpinMode = "Allowed"
	HairpinDisabled HairpinMode = "Disabled"
	// Necessary for backwards compatibility with HairpinMode
	HairpinUndefined HairpinMode = ""
)

```

twz123 commented on pkg/apis/k0s.k0sproject.io/v1beta1/kuberouter.go:
```diff
@@ -24,19 +24,32 @@ type KubeRouter struct {
 	MetricsPort int `json:"metricsPort"`
-	// Activate Hairpin Mode (allow a Pod behind a Service to communicate to its own ClusterIP:Port)
-	HairpinMode bool `json:"hairpinMode"`
+	// Admits three values: "enabled" enables it globaly, "allowed" allows but services must be annotated explictly and "disabled"
+	Hairpin Hairpin `json:"hairpin"`
+	// DEPRECATED: Use hairpining instead. Activates Hairpin Mode (allow a Pod behind a Service to communicate to its own ClusterIP:Port)
+	HairpinMode bool `json:"hairpinMode,omitempty"`
```
It's a pity that the better name is already taken by the old boolean setting. I wonder if we could merge those into one. Either accepting booeans or strings, just like apimachinery's `IntOrString` does it for ints and strings.

In any case I think we should rename both the type and the field name to `HairpinMode`, but just leave the json annotations untouched.

```suggestion
	// Admits three values: "Enabled" enables it globally, "Allowed" allows but services must be annotated explicitly and "Disabled"
	// +optional
	HairpinMode HairpinMode `json:"hairpin,omitempty"`
	// DEPRECATED: Use hairpining instead. Activates Hairpin Mode (allow a Pod behind a Service to communicate to its own ClusterIP:Port)
	HairpinEnabled bool `json:"hairpinMode,omitempty"`
```

jnummelin replied:
> Either accepting booleans or strings, just like apimachinery's IntOrString does it for ints and strings.
I like this idea very much

twz123 commented on pkg/component/controller/kuberouter.go:
```diff
@@ -70,6 +71,27 @@ func (k *KubeRouter) Init(_ context.Context) error { return nil }
 // Stop no-op as nothing running
 func (k *KubeRouter) Stop() error { return nil }
 
+func getHairpinConfig(cfg *kubeRouterConfig, krc *v1beta1.KubeRouter) {
```
Maybe a better name this `applyHairpinConfig`, since this is not idempotently getting something, but updating the cfg in place.

twz123 reviewed (commented):
We should probably implement `json.Unmarshaller` for `KubeRouter`, now that we have a default which doesn't equal the zero value of its type.

mtb-xt commented:
There was an edge case about the CNI config - if it was already generated, then activating hairpin mode in k0s won't update the CNI config. 
If someone now creates the cluster with hairpin mode disabled, but then decides to enable it - it won't be updated as well.
I realize this is a rare scenario now it's enabled by default, but do we want to do anything about it or not?

https://github.com/k0sproject/k0s/blob/78b4567f78abbff122a2cbf4ac88e1e51c737b9c/pkg/component/controller/kuberouter.go#L232 <- because of this check, if there is a CNI config present, it won't be touched.

juanluisvaladas commented:
Having one field accepting multiple types breaks dynamic configuration because of https://github.com/kubernetes-sigs/controller-tools/issues/735

I'm currently waiting for controller-manager people to give me the OK to start implementing this so in the meantime I'll put this on hold. If the proposed change wouldn't be accepted  then we'll have to consider a strategy different than using the same field to accept multiple types.

juanluisvaladas commented:
@jnummelin @twz123 I don't think I can achieve this even by patching the CRD manually.

The CRDs with `IntOrString` work because the porperties of the schema include the field `x-kubernetes-int-or-string`. The apiserver schema validation [explictly skips some validations for IntOrString](https://github.com/kubernetes/kubernetes/blob/v1.26.0/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go#L45-L62), so we cannot modify this.

I've tried to workaround this in many ways but I don't think it's actually possible. The only option that I see here would be adding x-kubernetes-preserve-unknown-fields to the KubeRouter struct and removing the HairpinMode field of the CRD but then this adds some additional problems in validation...

I see the following options:
1- Issue a new version of the clusterconfigs crd including a new version v1beta2 in which the hairpinMode is a string. This is probably the best one.
2- Add a new field with a different name and deprecate HairpinMode, which isn't great because HairpinMode is the best name.
3- Allowing unknown fields and removing the field from the CRD, which brings issues with the validation.

twz123 commented:
I would vote for option 2 (new field with a different name). This is what this PR initially aimed at, and it's a fair compromise compared to the alternatives. Introducing a new API version for this seems like a quite heavy thing to do. Maybe it's better to accumulate some changes over time and then issue a new API version that contains more improvements than just this one. Allowing unknown fields OTOH seems to me like removing a safety net for very little gain.

I didn't know that IntOrString was such a custom thing. Otherwise I wouldn't have brought it up as an example :see_no_evil:.

juanluisvaladas commented:
> I would vote for option 2 (new field with a different name).

And there is a big advantage that it's already implemented, I just had to go through the reflog :) and patch some little thing.

twz123 commented on pkg/apis/k0s.k0sproject.io/v1beta1/kuberouter.go:
```diff
@@ -24,19 +24,32 @@ type KubeRouter struct {
 	PeerRouterIPs string `json:"peerRouterIPs"`
 }
 
+type Hairpin string
```
I'd add an enum hint for kubebuilder.

```suggestion
// +kubebuilder:validation:Enum=Enabled;Allowed;Disabled
type Hairpin string
```

juanluisvaladas replied:
Ugh stupid cherry-pick. This was done. I need to review this more carefully.

twz123 commented on pkg/apis/k0s.k0sproject.io/v1beta1/kuberouter.go:
```diff
@@ -24,19 +24,32 @@ type KubeRouter struct {
 	MetricsPort int `json:"metricsPort"`
-	// Activate Hairpin Mode (allow a Pod behind a Service to communicate to its own ClusterIP:Port)
-	HairpinMode bool `json:"hairpinMode"`
+	// Admits three values: "enabled" enables it globaly, "allowed" allows but services must be annotated explictly and "disabled"
```
Some small corrections.

```suggestion
	// Admits three values: "Enabled" enables it globally, "Allowed" allows but services must be annotated explicitly and "Disabled".
```

twz123 commented on pkg/apis/k0s.k0sproject.io/v1beta1/kuberouter.go:
```diff
@@ -24,19 +24,32 @@ type KubeRouter struct {
-	// Activate Hairpin Mode (allow a Pod behind a Service to communicate to its own ClusterIP:Port)
-	HairpinMode bool `json:"hairpinMode"`
+	// Admits three values: "enabled" enables it globaly, "allowed" allows but services must be annotated explictly and "disabled"
+	Hairpin Hairpin `json:"hairpin"`
```
Worth mentioning the default value here as well.

```suggestion
	// Defaults to "Enabled".
	// +kubebuilder:default=Enabled
	Hairpin Hairpin `json:"hairpin"`
```

twz123 replied:
The kubebuilder default hint is still missing. But that's a purely cosmetic thing...

twz123 commented on docs/configuration.md:
```diff
@@ -66,7 +66,7 @@ spec:
       peerRouterASNs: ""
       autoMTU: true
-      hairpinMode: false
+      hairpin: enabled
```
```suggestion
      hairpin: Enabled
```

twz123 commented on pkg/apis/k0s.k0sproject.io/v1beta1/kuberouter.go:
```diff
@@ -24,19 +24,32 @@ type KubeRouter struct {
-	HairpinMode bool `json:"hairpinMode"`
+	// Admits three values: "enabled" enables it globaly, "allowed" allows but services must be annotated explictly and "disabled"
+	Hairpin Hairpin `json:"hairpin"`
+	// DEPRECATED: Use hairpining instead. Activates Hairpin Mode (allow a Pod behind a Service to communicate to its own ClusterIP:Port)
```
```suggestion
	// DEPRECATED: Use hairpin instead. Activates Hairpin Mode (allow a Pod behind a Service to communicate to its own ClusterIP:Port)
```

juanluisvaladas commented:
@twz123 I addressed everything and I also changed the table in docs/configuration.md. Apparently I didn´t find the latest version if the reflog and there were a couple things missing even if the code was working as expected.

I went file by file and I believe everything is good now.

juanluisvaladas commented:
Oh sorry I missed that one. That's added now.
